File: src/lib/attachments.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * Typed IPC wrappers for the per-feature attachment subsystem.
 *
 * Mirrors the shape of `src/lib/agentModels.ts` — never call `invoke()`
 * directly from a component (`AGENTS.md` §4). All commands correspond to
 * the `feature_add_attachment` / `feature_list_attachments` /
 * `feature_remove_attachment` / `attachment_read` Tauri handlers in
 * `src-tauri/src/commands/attachments.rs`.
 *
 * **Wire contract.** The Rust side stores attachments as content-addressable
 * blobs under `<app_local_data_dir>/attachments/<feature_id>/<sha256>.<ext>`
 * (see `domain::attachment::AttachedFile`). Re-uploading the same bytes
 * under a different filename is idempotent — the manifest row is updated
 * to reflect the new name but the on-disk file is shared.
 *
 * **Launch-stage model.** Because `feature_add_attachment` requires a
 * `feature_id` (the Rust side has no "staged / null feature" code path —
 * the spec's stage-then-claim model collapsed into per-feature attach in
 * sub-1), pre-launch composers ({@link StartFeatureModal},
 * {@link ProjectHome}) collect File handles + paths locally in a
 * `LaunchStaging` Map and call `addAttachment` only after the launching
 * feature id is known. This decouples UI feedback from feature creation
 * and keeps the chip list live during form fill-in.
 *
 * **Preview reads.** Post-launch, attachments fetched via
 * {@link listAttachments} arrive without a browser `File` handle (they
 * came in through Tauri drag-and-drop, which yields absolute paths
 * only). {@link readAttachment} round-trips the bytes back to the
 * webview through the `attachment_read` IPC so the preview Modal can
 * render out-of-session files. {@link getAttachmentDataUrl} is the
 * thin shim that turns those bytes into a `data:` URL — when a
 * browser `File` is available it goes through FileReader (instant,
 * no IPC); when only the manifest row is available it falls through
 * to {@link readAttachment} provided a `featureId` is supplied.
 */
export interface AttachedFile {
  /** Backend-generated stable id, format `at-<random>`. */
  id: string;
  /** Sanitized display name (used for chip labels). */
  name: string;
  /** IANA mime, e.g. `image/png`. */
  mime: string;
  /** Lowercase hex SHA-256 of the on-disk bytes. */
  sha256: string;
  size: number;
  /** Original user-supplied filename, preserved verbatim for the UI. */
  source_filename: string;
}

/**
 * A browser-file attachment retained until a newly launched Feature has an
 * id. This is intentionally frontend-only: callers persist these through the
 * existing `feature_add_attachment` flow after launch.
 */
export interface LaunchStageEntry {
  sha256: string;
  name: string;
  source_filename: string;
  mime: string;
  size: number;
  previewUrl: string | null;
  file: File | null;
  sourcePath: string | null;
}

/**
 * Input accepted by {@link addAttachment}. Either a browser `File`
 * (from `<input type="file">` or clipboard paste) OR an absolute local
 * path string (from Tauri's drag-drop event or `plugin-dialog`).
 *
 * The wrapper never reads bytes through this object — the Rust command
 * reads the file directly from disk via `std::fs::read`. Local `File`
 * handles are kept only so the UI can render a preview thumbnail.
 */
export type AttachmentInput =
  | { kind: "file"; file: File }
  | { kind: "path"; sourcePath: string; sourceFilename?: string; mime?: string };

/**
 * Add an attachment to a feature.
 *
 * Routes to the Tauri command `feature_add_attachment`. The Rust side
 * validates size (max 100 MiB), dedupes by content hash, and writes
 * the bytes under `<attachments_root>/<feature_id>/<sha256>.<ext>`.
 *
 * **Bytes vs path.** Modern Chromium / Tauri 2 webviews strip the
 * legacy `File.path` extension on `<input type="file">` selections
 * for security, so a click-picked browser `File` typically arrives
 * without an absolute disk path. When the input is a `file` kind
 * (drag-and-drop's "path" kind still has an absolute path), we read
 * the bytes into memory and ferry them through IPC as a JSON number
 * array — the Rust command accepts either a path or bytes. Drag-and-
 * drop paths continue to take the path branch unchanged.
 *
 * @param featureId target feature id (already-launched features only).
 * @param input     file handle (with preview-friendly FileReader) or
 *                  absolute path (drag-and-drop / native dialog).
 */
export async function addAttachment(
  featureId: string,
  input: AttachmentInput,
): Promise<AttachedFile> {
  const wire = await attachmentWire(input);
  return invoke<AttachedFile>("feature_add_attachment", {
    featureId,
    sourcePath: wire.sourcePath,
    mime: wire.mime,
    sourceFilename: wire.sourceFilename,
    bytes: wire.bytes,
  });
}

/** The four arguments every `*_add_attachment` command takes, whichever way
 *  the file was picked. Shared so a second owner (a Ticket, a Discovery)
 *  cannot decide the bytes-vs-path question differently from a Feature. */
export interface AttachmentWire {
  sourcePath: string;
  mime: string | null;
  sourceFilename: string | null;
  bytes: number[] | null;
}

/**
 * Normalise a pick into that shape.
 *
 * The browser gives a `File` but no usable absolute path (the common case in
 * Tauri 2 — `file.path` is stripped), so its bytes are read here and ferried
 * through IPC. Drag-and-drop yields a path and keeps the path branch, where
 * the Rust command reads the bytes itself.
 */
export async function attachmentWire(input: AttachmentInput): Promise<AttachmentWire> {
  if (input.kind === "file") {
    const bytes = new Uint8Array(await input.file.arrayBuffer());
    return {
      sourcePath: "",
      mime: input.file.type || null,
      sourceFilename: input.file.name,
      bytes: Array.from(bytes),
    };
  }
  return {
    sourcePath: input.sourcePath,
    mime: input.mime ?? null,
    sourceFilename: input.sourceFilename ?? pathBasename(input.sourcePath),
    bytes: null,
  };
}

/**
 * List every attachment on a feature. Returns `[]` when the feature has
 * no attachments column populated (the manifest column default).
 */
export async function listAttachments(featureId: string): Promise<AttachedFile[]> {
  const list = await invoke<AttachedFile[]>("feature_list_attachments", {
    featureId,
  });
  return Array.isArray(list) ? list : [];
}

/**
 * Remove an attachment. Idempotent: the Rust side is a no-op if the id
 * is already gone. The on-disk file is shared by content hash, so it
 * is deleted only when no other manifest row references the same sha256.
 */
export async function removeAttachment(
  featureId: string,
  attachmentId: string,
): Promise<void> {
  await invoke<void>("feature_remove_attachment", {
    featureId,
    attachmentId,
  });
}

/**
 * Result of {@link readAttachment}: the on-disk mime plus the raw bytes.
 *
 * The Rust command serializes `Vec<u8>` as a JSON array of numbers
 * (0-255), which we repack into a `Uint8Array` on the JS side. For
 * preview-only display paths (e.g. an `<img src="data:...">` Modal).
 *
 * Never used on the prompt-injection path — the orchestrator mirrors
 * bytes into the per-step worktree on the Rust side, not via IPC.
 */
export interface AttachmentBytes {
  mime: string;
  bytes: Uint8Array;
}

/**
 * Fetch the bytes of a previously-uploaded attachment via the
 * `attachment_read` IPC.
 *
 * Use case: a preview Modal needs to render an out-of-session file
 * (one that arrived through Tauri drag-and-drop with no browser
 * `File` handle). Resolves the manifest row server-side, so callers
 * don't need to pass `mime` or `sha256` — they get both back.
 *
 * Throws when the feature or attachment id is not present in the
 * manifest, when the feature is missing on disk, or when the
 * underlying bytes can't be read.
 */
export async function readAttachment(
  featureId: string,
  attachmentId: string,
): Promise<AttachmentBytes> {
  const raw = await invoke<number[]>("attachment_read", {
    featureId,
    attachmentId,
  });
  const manifest = await listAttachments(featureId);
  const meta = manifest.find((a) => a.id === attachmentId);
  return {
    mime: meta?.mime ?? "application/octet-stream",
    bytes: Uint8Array.from(raw),
  };
}

/**
 * Generate a `data:<mime>;base64,…` URL for a picked file, or null
 * when no source can produce bytes.
 *
 * Used by the chip preview + hover-preview surfacing in
 * {@link AttachmentChip} / {@link AttachmentDropzone}. Three modes:
 *
 * 1. A browser `File` is supplied → FileReader path (instant, no IPC).
 * 2. A `featureId` is supplied → falls through to
 *    {@link readAttachment} via the `attachment_read` IPC. This is the
 *    post-launch preview path for files that came in through Tauri
 *    drag-and-drop (no browser File handle).
 * 3. Neither → returns null. Pre-launch callers that haven't supplied
 *    a File yet fall back to a mime-icon chip; see {@link AttachmentChip}.
 */
export async function getAttachmentDataUrl(
  attachment: AttachedFile,
  file?: File | null,
  featureId?: string,
): Promise<string | null> {
  if (file) {
    return readFileAsDataUrl(file);
  }
  if (featureId) {
    const { mime, bytes } = await readAttachment(featureId, attachment.id);
    return bytesToDataUrl(mime, bytes);
  }
  return null;
}

function bytesToDataUrl(mime: string, bytes: Uint8Array): string {
  // Chunked String.fromCharCode avoids a "Maximum call stack size
  // exceeded" on the larger file cap (25 MiB) where a single
  // fromCharCode(bytes) would blow the JS argument limit.
  let binary = "";
  const CHUNK = 0x8000;
  for (let i = 0; i < bytes.length; i += CHUNK) {
    const slice = bytes.subarray(i, Math.min(i + CHUNK, bytes.length));
    binary += String.fromCharCode.apply(null, Array.from(slice));
  }
  const b64 = btoa(binary);
  return `data:${mime};base64,${b64}`;
}

/**
 * Exact allow-list of clipboard image MIME types that the rest of the
 * pipeline accepts. Kept in sync with the Rust
 * `commit_attachment_inner` allow-list at
 * `crates/demeteo-core/src/application/attachments.rs:221-235`; a
 * clipboard item outside this set never reaches `addAttachment` and
 * the underlying `feature_add_attachment` would reject it anyway.
 */
const SUPPORTED_CLIPBOARD_IMAGE_MIMES: ReadonlySet<string> = new Set([
  "image/png",
  "image/jpeg",
  "image/gif",
  "image/webp",
  "image/tiff",
]);

/**
 * The browser-visible outcome of inspecting clipboard file entries for
 * supported images. `unavailable` means a supported image was advertised,
 * but its bytes cannot be obtained through the webview's `File` API.
 */
export type ClipboardImageExtraction =
  | { kind: "files"; files: File[] }
  | { kind: "none" }
  | { kind: "unavailable"; mime: string };

/**
 * Result of the async Clipboard API recovery path. This is deliberately
 * separate from {@link ClipboardImageExtraction}: callers use it only when a
 * paste event exposed no items at all (the WebKitGTK 218519 shape).
 */
export type AsyncClipboardImageRecovery =
  | { kind: "recovered"; file: File }
  | { kind: "unavailable" }
  | { kind: "denied" }
  | { kind: "no-bytes" };

/**
 * Read one supported clipboard image directly from the async Clipboard API.
 *
 * This must be called from the paste gesture, after the synchronous
 * DataTransfer path found no image. It does not broaden the MIME allow-list:
 * only types in {@link SUPPORTED_CLIPBOARD_IMAGE_MIMES} are requested.
 */
export async function recoverClipboardImageFile(): Promise<AsyncClipboardImageRecovery> {
  const clipboard = navigator.clipboard;
  if (!clipboard?.read) return { kind: "unavailable" };

  let items: ClipboardItem[];
  try {
    items = await clipboard.read();
  } catch {
    return { kind: "denied" };
  }

  for (const item of items) {
    for (const advertisedType of item.types) {
      const mime = advertisedType.toLowerCase();
      if (!SUPPORTED_CLIPBOARD_IMAGE_MIMES.has(mime)) continue;
      try {
        const blob = await item.getType(advertisedType);
        if (blob.size === 0) continue;
        const extension = mime === "image/jpeg" ? "jpg" : mime.slice("image/".length);
        return {
          kind: "recovered",
          file: new File([blob], `pasted-image.${extension}`, { type: mime }),
        };
      } catch {
        // Another clipboard item may still expose usable bytes.
      }
    }
  }

  return { kind: "no-bytes" };
}

/**
 * Extract supported clipboard images while preserving the distinction between
 * no supported image and a supported image the browser cannot expose as a
 * `File`. MIME matching and unavailable MIME reporting are lowercase.
 *
 * Text, HTML, and unsupported entries are deliberately not inspected: this
 * helper only invokes `getAsFile()` after an item passes the existing file
 * kind and MIME allow-list checks.
 */
export function extractClipboardImageFiles(
  clipboardData: DataTransfer,
): ClipboardImageExtraction {
  const files: File[] = [];
  const items = clipboardData.items;
  for (let i = 0; i < items.length; i++) {
    const item = items[i];
    if (item.kind !== "file") continue;

    const mime = item.type.toLowerCase();
    if (!SUPPORTED_CLIPBOARD_IMAGE_MIMES.has(mime)) continue;

    const file = item.getAsFile();
    if (file === null) return { kind: "unavailable", mime };
    files.push(file);
  }

  return files.length > 0 ? { kind: "files", files } : { kind: "none" };
}

/**
 * Extract the supported image `File` handles from a clipboard /
 * drag-paste `DataTransfer`. Returns the items in clipboard order;
 * `[]` when no supported image is present.
 *
 * Only entries whose `kind === "file"` and whose `type` matches the
 * exact allow-list of supported image MIME types (compared
 * case-insensitively) are considered. `getAsFile()` is invoked only
 * for items that pass both filters. A supported item whose file is
 * unavailable maps to an empty compatibility result; callers needing to
 * distinguish it from no image should use {@link extractClipboardImageFiles}.
 *
 * Pure: no I/O, no IPC, no `preventDefault`, no filename normalization
 * or hashing — those happen later in `ingestFiles` /
 * `feature_add_attachment`. The caller (e.g.
 * `AttachmentDropzone.handlePaste`) decides whether to swallow the
 * event based on whether this helper returned a non-empty list.
 */
export function extractImageFilesFromClipboard(
  clipboardData: DataTransfer,
): File[] {
  const extraction = extractClipboardImageFiles(clipboardData);
  return extraction.kind === "files" ? extraction.files : [];
}

/**
 * Compute SHA-256 hex over a browser `File` using `crypto.subtle`.
 * Used by the launch-stage dedup in {@link AttachmentDropzone} so the
 * staging Map keys the same way the Rust `feature_add_attachment`
 * command keys the manifest (sha256). Returns lowercase hex (64 chars).
 */
export async function computeLocalSha256(file: File): Promise<string> {
  const buf = await file.arrayBuffer();
  const digest = await crypto.subtle.digest("SHA-256", buf);
  return Array.from(new Uint8Array(digest))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

/**
 * Build launch-stage entries for browser `File` values and replace prior
 * entries with matching byte hashes. All browser-file launch entry points
 * (picker, paste, modal, and composer) share this so their MIME fallback,
 * preview generation, and SHA-256 deduplication cannot drift.
 */
export async function stageBrowserFilesForLaunch(
  files: readonly File[],
  stageEntries: readonly LaunchStageEntry[],
): Promise<LaunchStageEntry[]> {
  let next = [...stageEntries];

  for (const file of files) {
    const sourceFilename = file.name;
    const mime = file.type.toLowerCase() || guessBrowserFileMime(sourceFilename.toLowerCase());
    const sha256 = await computeLocalSha256(file);
    const previewUrl = mime.startsWith("image/") ? await readFileAsDataUrl(file) : null;
    const entry: LaunchStageEntry = {
      sha256,
      name: sourceFilename,
      source_filename: sourceFilename,
      mime,
      size: file.size,
      previewUrl,
      file,
      sourcePath: null,
    };
    next = [...next.filter((existing) => existing.sha256 !== sha256), entry];
  }

  return next;
}

/**
 * One file collected before the aggregate that will own it exists — the wire
 * shape `start_feature`, `remote_submit_run` and `discovery_create` all take.
 */
export interface StagedAttachmentInput {
  source_path: string;
  mime: string | null;
  source_filename: string | null;
  bytes: number[] | null;
}

/**
 * Convert a staged batch into that shape.
 *
 * The whole batch travels with the create call rather than as a follow-up per
 * file, which is what stops the agent's first turn from racing the attachments
 * it is supposed to have been given.
 */
export async function stagedAttachmentInputs(
  entries: readonly LaunchStageEntry[],
): Promise<StagedAttachmentInput[]> {
  return Promise.all(
    entries.map(async (entry) => ({
      source_path: entry.sourcePath ?? "",
      mime: entry.mime ?? null,
      source_filename: entry.source_filename ?? null,
      bytes: entry.file ? Array.from(new Uint8Array(await entry.file.arrayBuffer())) : null,
    })),
  );
}

/**
 * Stage-time metadata for a path-based attachment pick (Tauri
 * drag-and-drop yields an absolute path; no browser `File` is
 * available to Web Crypto).
 *
 * Routes to the Tauri command `attachment_stage_metadata`. The Rust
 * side reads the bytes from `sourcePath`, sha256s them, and returns
 * the size — mirroring the bytes-fetch + sha256 surface that
 * `feature_add_attachment` produces, minus the feature-scoped
 * storage step (no `feature_id` exists at staging time). The React
 * launch-stage uses `sha256` as both the React key AND the
 * re-drop dedup signal, and `size` so the chip renders the real
 * byte count instead of a confusing "0 B".
 *
 * Errors propagate from Rust as `AppError::validation` (missing
 * file, oversize, unsupported mime/ext) so the dropzone surfaces
 * an inline error instead of a silently-staged entry.
 */
export interface StagedAttachmentMeta {
  /** Lowercase hex SHA-256 of the on-disk bytes (matches the
   *  `feature_add_attachment` server-side computation byte-for-byte). */
  sha256: string;
  /** Byte length, used to render the chip's size label. */
  size: number;
}

/**
 * Fetch the staging-time metadata for a dropped file.
 *
 * @param sourcePath absolute disk path from Tauri drag-and-drop.
 * @param mime       optional mime hint (frontend can infer from filename).
 * @param sourceFilename optional original filename (used to infer mime).
 */
export async function stageAttachmentMetadata(
  sourcePath: string,
  mime?: string | null,
  sourceFilename?: string | null,
): Promise<StagedAttachmentMeta> {
  return invoke<StagedAttachmentMeta>("attachment_stage_metadata", {
    sourcePath,
    mime: mime ?? null,
    sourceFilename: sourceFilename ?? null,
  });
}

function readFileAsDataUrl(file: File): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onerror = () => reject(reader.error ?? new Error("FileReader failed"));
    reader.onload = () => {
      const out = reader.result;
      if (typeof out === "string") resolve(out);
      else reject(new Error("FileReader did not yield a string result"));
    };
    reader.readAsDataURL(file);
  });
}

function guessBrowserFileMime(lowerFilename: string): string {
  if (lowerFilename.endsWith(".png")) return "image/png";
  if (lowerFilename.endsWith(".jpg") || lowerFilename.endsWith(".jpeg")) return "image/jpeg";
  if (lowerFilename.endsWith(".gif")) return "image/gif";
  if (lowerFilename.endsWith(".webp")) return "image/webp";
  if (lowerFilename.endsWith(".tif") || lowerFilename.endsWith(".tiff")) return "image/tiff";
  if (lowerFilename.endsWith(".pdf")) return "application/pdf";
  if (lowerFilename.endsWith(".md") || lowerFilename.endsWith(".markdown")) return "text/markdown";
  if (lowerFilename.endsWith(".txt")) return "text/plain";
  if (lowerFilename.endsWith(".json")) return "application/json";
  return "application/octet-stream";
}

/**
 * Split a path on either separator and return the trailing segment.
 * Used by the drag-and-drop "path" branch where we already have an
 * absolute disk path and only need the filename for the manifest row.
 */
function pathBasename(p: string): string {
  const parts = p.split(/[\\/]/).filter(Boolean);
  return parts.length === 0 ? p : parts[parts.length - 1];
}

File: src-tauri/src/commands/attachments.rs
//! Tauri commands for the per-feature attachment subsystem.
//!
//! Four additive commands wired through `commands::mod`:
//!
//! * `feature_add_attachment` — validate (size, ext, sha256), copy
//!   bytes from a user-supplied absolute path into the
//!   `AttachmentStore`, dedup by content hash, append an
//!   [`AttachedFile`] row to the feature's manifest via
//!   [`AttachmentJsonPort`], and return the new manifest entry. A
//!   re-upload of the same content under a different filename is an
//!   idempotent no-op beyond updating the row metadata to reflect
//!   the latest name — the on-disk file is shared.
//! * `feature_list_attachments` — return the feature's manifest, or
//!   `[]` if the feature has no attachments column populated.
//! * `feature_remove_attachment` — drop the manifest entry and the
//!   on-disk bytes. Idempotent.
//! * `attachment_read` — return the bytes of a previously-uploaded
//!   attachment for the React preview Modal. Resolves the row by
//!   `(feature_id, attachment_id)` and goes through the same
//!   path-within-root check as every other on-disk read. Never used
//!   on the prompt-injection path.
//!
//! Validation rules (mirrored in the Start-Feature modal and Gate
//! view):
//!
//! * Reject if the file does not exist or is not a regular file.
//! * Reject if `size > 100 MiB` (v1 hard cap from the implementation
//!   spec).
//! * Reject if the per-feature attachment count would exceed 10.
//! * Refuse a re-upload whose bytes don't match an existing
//!   attachment's sha256 (the on-disk path would be `<sha256>.<ext>`,
//!   so two different contents can never share a path). The
//!   `FsAttachmentStore::write` defensive collision check enforces
//!   this and surfaces the error here.
//!
//! No new feature keys or capabilities — file reads happen inside
//! Rust, outside the webview's `fs:` scope.

use crate::domain::attachment::{compute_sha256_hex, ext_for_mime, AttachedFile};
use crate::domain::ids::FeatureId;
use crate::error::AppError;
use crate::state::AppContext;
use demeteo_core::application::attachments::{
    commit_attachment_inner, is_supported_attachment, resolve_mime, MAX_ATTACHMENT_BYTES,
};
use serde::Serialize;
use std::path::Path;
use tauri::State;
use tracing::{info, warn};

/// Minimal metadata about a file on disk that the launch-stage
/// (`AttachmentDropzone` `mode === "launch"`) needs BEFORE a feature
/// exists: SHA-256 for the local dedup key (so re-dropping the same
/// path collapses to one chip) and the byte length (so the chip
/// renders the real size instead of a confusing "0 B").
///
/// Returned by the [`attachment_stage_metadata`] Tauri command.
/// Mirrors the bytes + sha256 surface that `feature_add_attachment`
/// produces server-side, minus the feature-scoped storage step (no
/// feature exists yet at staging time).
#[derive(Debug, Clone, Serialize)]
pub struct StagedAttachmentMeta {
    pub sha256: String,
    pub size: u64,
}

/// Staged attachment supplied at feature-start time. Business logic lives
/// in `demeteo_core::application::attachments` (shared with the step
/// executor's pre-execution path) — re-exported here so existing
/// `commands::attachments::StagedAttachmentInput` call sites keep working.
pub use demeteo_core::application::attachments::StagedAttachmentInput;

/// Add an attachment to a feature.
///
/// `bytes` carries the in-memory attachment bytes when the caller
/// has a browser `File` handle but no absolute path on disk — modern
/// Chromium / Tauri 2 webviews strip the legacy `File.path`
/// extension on `<input type="file">` selections for security, so
/// the only way to ferry the bytes is through IPC. Serialized as
/// `number[]` (JSON array of 0–255 ints) for cross-platform
/// compatibility — mirrors the return shape of `attachment_read`.
/// When `Some`, `source_path` is ignored.
#[tauri::command]
pub async fn feature_add_attachment(
    ctx: State<'_, AppContext>,
    feature_id: String,
    source_path: String,
    mime: Option<String>,
    source_filename: Option<String>,
    bytes: Option<Vec<u8>>,
) -> Result<AttachedFile, AppError> {
    commit_attachment_inner(
        &ctx.features,
        &ctx.attachment_json,
        &ctx.attachments,
        &feature_id,
        &source_path,
        mime.as_deref(),
        source_filename.as_deref(),
        bytes,
    )
}

/// Compute the staging-time metadata for a path-based pick (Tauri
/// drag-and-drop yields an absolute path, no browser `File`).
///
/// Reads the file once, returns `{ sha256, size }`. The React
/// launch-stage uses the sha256 as the chip's React key AND as the
/// dedup signal (so re-dropping the same path collapses to one chip)
/// and the size to render the chip's byte-count label. Behaviour
/// matches [`commit_attachment_inner`] for the bytes-fetch + sha256
/// + support-check steps; the storage + manifest write are not
/// performed (no feature_id exists yet at staging time).
///
/// Errors mirror [`commit_attachment_inner`]: missing file, oversized
/// file, unsupported mime/ext. Returning `AppError` means the
/// dropzone surfaces an inline error to the user instead of a
/// silently-staged entry.
#[tauri::command]
pub async fn attachment_stage_metadata(
    source_path: String,
    mime: Option<String>,
    source_filename: Option<String>,
) -> Result<StagedAttachmentMeta, AppError> {
    let src = std::path::PathBuf::from(&source_path);
    let meta = std::fs::metadata(&src).map_err(|e| {
        AppError::validation(format!("could not stat source file {}: {}", source_path, e))
    })?;
    if !meta.is_file() {
        return Err(AppError::validation(format!(
            "source path is not a regular file: {}",
            source_path
        )));
    }
    if meta.len() > MAX_ATTACHMENT_BYTES {
        return Err(AppError::validation(format!(
            "attachment too large: {} bytes (max {})",
            meta.len(),
            MAX_ATTACHMENT_BYTES
        )));
    }
    let bytes = std::fs::read(&src).map_err(|e| {
        AppError::validation(format!("could not read source file {}: {}", source_path, e))
    })?;
    let sha256 = compute_sha256_hex(&bytes);
    let resolved_mime = resolve_mime(mime.as_deref(), source_filename.as_deref(), &src);
    let ext = ext_for_mime(&resolved_mime)
        .map(str::to_string)
        .unwrap_or_else(|| {
            Path::new(source_filename.as_deref().unwrap_or(&source_path))
                .extension()
                .and_then(|s| s.to_str())
                .map(|s| s.to_ascii_lowercase())
                .unwrap_or_else(|| "bin".to_string())
        });
    if !is_supported_attachment(&resolved_mime, &ext) {
        return Err(AppError::validation(format!(
            "unsupported attachment type: mime={} ext={} (allowed: png, jpg, gif, webp, pdf, txt, md, json)",
            resolved_mime, ext
        )));
    }
    Ok(StagedAttachmentMeta {
        sha256,
        size: bytes.len() as u64,
    })
}

#[tauri::command]
pub async fn feature_list_attachments(
    ctx: State<'_, AppContext>,
    feature_id: String,
) -> Result<Vec<AttachedFile>, AppError> {
    let fid = FeatureId::from(feature_id);
    let _ = ctx
        .features
        .get(&fid)?
        .ok_or_else(|| AppError::not_found(format!("feature not found: {}", fid.as_str())))?;
    ctx.attachment_json
        .get_attachments(&fid)
        .map_err(AppError::from)
}

/// Read the bytes of a previously-uploaded attachment.
///
/// Resolves the manifest row by `attachment_id` (scoped to the feature
/// so an attacker can't probe other features' attachments by guessing
/// ids), derives the on-disk extension the same way
/// [`feature_remove_attachment`] does, and returns the bytes via the
/// existing [`AttachmentStore::read`] port — which enforces the
/// "path within attachments root" safety check before touching the
/// filesystem.
///
/// Use case: the React preview Modal for out-of-session files (files
/// that arrived through Tauri drag-and-drop with no browser `File`
/// handle). Never used on the prompt-injection path — the orchestrator
/// already mirrors bytes into the per-step worktree via
/// `resolve_and_materialize_attachments`.
#[tauri::command]
pub async fn attachment_read(
    ctx: State<'_, AppContext>,
    feature_id: String,
    attachment_id: String,
) -> Result<Vec<u8>, AppError> {
    let fid = FeatureId::from(feature_id.clone());
    let _ = ctx
        .features
        .get(&fid)?
        .ok_or_else(|| AppError::not_found(format!("feature not found: {}", feature_id)))?;

    let current = ctx
        .attachment_json
        .get_attachments(&fid)
        .map_err(AppError::from)?;

    let attached = current
        .iter()
        .find(|a| a.id == attachment_id)
        .cloned()
        .ok_or_else(|| {
            AppError::not_found(format!(
                "attachment {} not found on feature {}",
                attachment_id, feature_id
            ))
        })?;

    let ext = derive_ext(&attached.mime, &attached.source_filename);
    let path = ctx
        .attachments
        .lookup_path(&feature_id, &attached.sha256, &ext);
    let path_str = path.to_string_lossy().to_string();

    let bytes = ctx.attachments.read(&path_str).map_err(AppError::from)?;
    Ok(bytes)
}

#[tauri::command]
pub async fn feature_remove_attachment(
    ctx: State<'_, AppContext>,
    feature_id: String,
    attachment_id: String,
) -> Result<(), AppError> {
    let fid = FeatureId::from(feature_id.clone());
    let _ = ctx
        .features
        .get(&fid)?
        .ok_or_else(|| AppError::not_found(format!("feature not found: {}", feature_id)))?;

    let current = ctx.attachment_json.get_attachments(&fid)?;
    let mut remaining: Vec<AttachedFile> = Vec::with_capacity(current.len());
    let mut removed: Option<AttachedFile> = None;
    for a in current.into_iter() {
        if a.id == attachment_id {
            removed = Some(a);
        } else {
            remaining.push(a);
        }
    }

    let removed = match removed {
        Some(r) => r,
        None => return Ok(()), // idempotent: nothing to remove
    };

    ctx.attachment_json
        .set_attachments(&fid, &remaining)
        .map_err(AppError::from)?;

    // Best-effort on-disk cleanup; the bytes may already be shared
    // by another manifest row with the same sha256. If no other row
    // references this sha256, drop the file.
    let still_used = remaining.iter().any(|a| a.sha256 == removed.sha256);
    if !still_used {
        let ext = derive_ext(&removed.mime, &removed.source_filename);
        let path = ctx
            .attachments
            .lookup_path(&feature_id, &removed.sha256, &ext);
        if path.exists() {
            if let Err(e) = ctx.attachments.delete(&path.to_string_lossy()) {
                warn!(
                    error = %e,
                    path = %path.display(),
                    "could not delete attachment file (already absent?)"
                );
            }
        }
    }

    info!(
        feature_id = %feature_id,
        attachment_id = %attachment_id,
        sha256 = %removed.sha256,
        "feature attachment removed"
    );
    Ok(())
}

/// Lowercase extension for a stored attachment: prefer the mime
/// reverse-lookup, fall back to `source_filename`'s tail, then
/// `bin`. Mirrors the `feature_add_attachment` extension choice so
/// read/lookup/remove all hit the same `<sha256>.<ext>` path.
fn derive_ext(mime: &str, source_filename: &str) -> String {
    ext_for_mime(mime).map(str::to_string).unwrap_or_else(|| {
        Path::new(source_filename)
            .extension()
            .and_then(|s| s.to_str())
            .map(|s| s.to_ascii_lowercase())
            .unwrap_or_else(|| "bin".to_string())
    })
}

#[cfg(test)]
#[path = "../../tests/infrastructure/attachments_command.rs"]
mod tests;
